### PR TITLE
changed diversity input check

### PR DESCRIPTION
**Pull request by ndegunst**
I changed around the input check for the index as suggested in #519 . Sorry for the headache caused by #514.

### CODE REVIEW

**Comment by TuomasBorman** [comment]
No problem at all, it is maintainers job to evaluate the code, and I don't assume that you know yet all the things in mia. The main goal is to develop good things with the best possible quality.

1. Can you modify also TreeSE method (same thing there, but now with additional faith index)
2. Run BiocCheck::BiocCheck() (Check that the code added by you passes Bioconductor recommendations. There are warnings related to other lines, skip those.)
3. Run manual tests that the code does what it should do. (Can you run the code from the issue? Can you print the code and output here?)

File: R/estimateDiversity.R [hunk 1]
```diff
@@ -254,7 +254,11 @@ setMethod("estimateDiversity", signature = c(x="SummarizedExperiment"),
         }
 
         # input check
-        index<- match.arg(index, several.ok = TRUE)
+        supported_index <- c("coverage", "fisher", "gini_simpson", 
+                             "inverse_simpson", "log_modulo_skewness", "shannon")
+        if( !all( index %in% supported_index ) ){
+            stop("'index' contains unsupported indices.", call. = FALSE)
+        }
```
Comment (TuomasBorman) [inline R/estimateDiversity.R]: Looks good! Also check that the length of index is over 0 (it is not checked currently. NULL might cause problems --> test `index = NULL`

**Review by TuomasBorman — CHANGES_REQUESTED** [review]

**Comment by ndegunst** [comment]
Thanks for the tips !

I ran the BiocCheck with no warnings for the lines I modified and I ran some manual tests on top of the already existing tests, here are the results :

```R

tse <- estimateDiversity(tse, 
                              assay.type = "counts",
                              index = "test", 
                              name = "shannon")

```

Output : 
```
Error: 'index' contains unsupported indices.

```

```R
tse <- estimateDiversity(tse, 
                         assay.type = "counts",
                         name = "shannon")
```

Output : 
```
Error: 'name' must be a non-empty character value and have the same length as 'index'.

```

```R

tse <- estimateDiversity(tse, assay.type = "counts"
                         , index = NULL,
                         name = "test")

```

Output : 
```
Error in .local(x, assay.type, assay_name, index, name, ...) : 
  No 'index' specified.

```

Would you rather have it so that it throws an index error or a name error ?

File: R/estimateDiversity.R [hunk 2]
```diff
@@ -283,19 +291,31 @@ setMethod("estimateDiversity", signature = c(x="TreeSummarizedExperiment"),
             name = index, tree_name = "phylo", 
             ..., BPPARAM = SerialParam()){
         # input check
+        supported_index <- c("coverage", "fisher", "gini_simpson", 
+                             "inverse_simpson", "log_modulo_skewness",
+                             "shannon","faith")
+        
+        if(!length(index)>0){
+            stop("No 'index' specified.")
```
Comment (TuomasBorman) [inline R/estimateDiversity.R]: Remember to add call. = FALSE (test what it does)

File: R/estimateDiversity.R [hunk 3]
```diff
@@ -283,19 +291,31 @@ setMethod("estimateDiversity", signature = c(x="TreeSummarizedExperiment"),
...
+        }
```
Comment (TuomasBorman) [inline R/estimateDiversity.R]: Also you could remove this check and modify the test below

File: R/estimateDiversity.R [hunk 4]
```diff
@@ -283,19 +291,31 @@ setMethod("estimateDiversity", signature = c(x="TreeSummarizedExperiment"),
...
+        
+        if( !all( index %in% supported_index ) ){
```
Comment (TuomasBorman) [inline R/estimateDiversity.R]: This check could test

1. length > 0
2. All values are supported

The error message could be:

"'index' must be a character value specifying alpha diversity index to be calculated. Choose from the following options: ..."

In ... you could print supported types. Something like thia

`stop("'", paste0(supported_types, collapse = "', '"), "'", call. = FALSE)`

**Comment by TuomasBorman** [comment]
Looks good! Run `BiocCheck::BiocCheck()` and fix everything related to these added lines. (There are some other things related to other lines, but let's keep them unchanged at this point)

**Comment by ndegunst** [comment]
That should do it, we shouldn't use collapse within a paste call and if we do past it shouldn't be within a stop call. Therefore, I assembled the string before using it in the stop function.

**Comment by ndegunst** [comment]
It should be good now, bioconductor doesn't allow a paste statement within a stop so i created a new variable that pastes the allowed indices before passing it to the stop function.

File: R/estimateDiversity.R [hunk 5]
```diff
@@ -283,19 +288,26 @@ setMethod("estimateDiversity", signature = c(x="TreeSummarizedExperiment"),
             name = index, tree_name = "phylo", 
             ..., BPPARAM = SerialParam()){
         # input check
+        supported_index <- c("coverage", "fisher", "gini_simpson", "faith",
+                             "inverse_simpson", "log_modulo_skewness", "shannon")
+        index_string <- paste0("'", paste0(supported_index, collapse = "', '"), "'")
+        if ( !all(index %in% supported_index) || length(index) == 0) {
+            stop(paste("Unsupported index:", index_string), call. = FALSE)
```
Comment (TuomasBorman) [inline R/estimateDiversity.R]: The error message could still be more informative.

"'index' must be from the following options:"

File: R/estimateDiversity.R [hunk 6]
```diff
@@ -283,19 +288,26 @@ setMethod("estimateDiversity", signature = c(x="TreeSummarizedExperiment"),
...
                  call. = FALSE)
         }
         if (!is.null(assay_name)) {
             .Deprecated(old="assay_name", new="assay.type", "Now assay_name is deprecated. Use assay.type instead.")
-        }	
-        # Check indices
-        index <- match.arg(index, several.ok = TRUE)
+        }
+        
+
```
Comment (TuomasBorman) [inline R/estimateDiversity.R]: Remove extra line (one is enough)

With these, good to go

**Review by TuomasBorman — APPROVED** [review]

**Pull request merged**